src/brevitas/graph/gpxq.py

```diff
@@ -0,0 +1,243 @@
+            for parallel_layers in self.group_of_parallel_layers:
+                for name in parallel_layers:
+                    if name not in dict_of_layers:
+                        raise ValueError(
+                            "The layer {} is not present in the model or it is not supported for GPTQ"
+                            .format(name))
+                    del dict_of_layers[name]
+                names = '_'.join(parallel_layers)
+                dict_of_layers[names] = [
+                    (name, attrgetter(name)(self.model)) for name in parallel_layers]
+
+        # Print warning if hooks are attached to any module, since the normal forward flow of the
+        # network is highly disrupted during GPxQ
+        for _, parallel_layers in dict_of_layers.items():
+            for name, module in parallel_layers:
+                if len(module._forward_hooks) > 0 or len(module._forward_pre_hooks):
+                    warnings.warn(
+                        f'Hooks detected during setup for GPxQ. '
+                        f'Behaviour might deviate from what expected.')
+
+                # Attach hooks for GPTQ
+                if self._is_module_supported(module):
+                    gpxq = self.class_implementation(
+                        module, name, act_order=self.act_order, parallel_layers=parallel_layers)
+                    hook_fn = partial(gpxq.update_batch, current_layer=self.current_layer)
+                    self.hook_dict[name] = module.register_forward_pre_hook(hook_fn)
+                    self.gpxq_layers[name] = gpxq
+        if not self.use_quant_activations:
+            self.disable_quant_inference.disable_act_quantization(
+                self.model, is_training=self.model.training)
+            self.disable_quant_inference.disable_bias_quantization(
+                self.model, is_training=self.model.training)
+
+        self.num_layers = len(dict_of_layers)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.model.forward = self.orig_forward
+        if not self.use_quant_activations:
+            self.disable_quant_inference.enable_act_quantization(
+                self.model, is_training=self.model.training)
+            self.disable_quant_inference.enable_bias_quantization(
+                self.model, is_training=self.model.training)
+
+    def update(self):
+        for name in self.current_layer.layer_names:
+            self.gpxq_layers[name].single_layer_update()
+            self.hook_dict[name].remove()
+        self.current_layer.layer_names.clear()
+
+    @abstractmethod
+    def catch_stopfwd(self, *args, **kwargs):
+        pass
+
+
+class GPxQ(ABC):
+
+    def __init__(self, layer, name, act_order, parallel_layers=1) -> None:
+        self.layer = layer
+        self.name = name
+        self.act_order = act_order
+
+        weight = layer.weight.data
+        self.layer.weight_orig = deepcopy(layer.weight)
```

Comment:
Add a flag to control this creation. Do a buffer that is initialized with detach().clone() and put it under a if that checks that it doesn't exist yet.